sql/core/src/test/scala/org/apache/spark/sql/DataFrameTimeWindowingSuite.scala

```diff
@@ -490,4 +490,81 @@ class DataFrameTimeWindowingSuite extends QueryTest with SharedSparkSession {
       assert(attributeReference.dataType == tuple._2)
     }
   }
+
+  test("No need to filter data when the sliding window length is not redundant") {
+    // check the value column
+    val df1 = Seq(
+      ("2022-02-15 19:39:34", 1, "a"),
+      ("2022-02-15 19:39:56", 2, "a"),
+      ("2022-02-15 19:39:27", 4, "b")).toDF("time", "value", "id")
+      .select(window($"time", "9 seconds", "3 seconds", "0 second"), $"value")
+      .orderBy($"window.start".asc, $"value".desc).select("value")
+    val df2 = Seq(
+      (LocalDateTime.parse("2022-02-15T19:39:34"), 1, "a"),
+      (LocalDateTime.parse("2022-02-15T19:39:56"), 2, "a"),
+      (LocalDateTime.parse("2022-02-15T19:39:27"), 4, "b")).toDF("time", "value", "id")
+      .select(window($"time", "9 seconds", "3 seconds", "0 second"), $"value")
+      .orderBy($"window.start".asc, $"value".desc).select("value")
+
+    val df3 = Seq(
+      ("2022-02-15 19:39:34", 1, "a"),
+      ("2022-02-15 19:39:56", 2, "a"),
+      ("2022-02-15 19:39:27", 4, "b")).toDF("time", "value", "id")
+      .select(window($"time", "9 seconds", "3 seconds", "-2 second"), $"value")
+      .orderBy($"window.start".asc, $"value".desc).select("value")
+    val df4 = Seq(
+      (LocalDateTime.parse("2022-02-15T19:39:34"), 1, "a"),
+      (LocalDateTime.parse("2022-02-15T19:39:56"), 2, "a"),
+      (LocalDateTime.parse("2022-02-15T19:39:27"), 4, "b")).toDF("time", "value", "id")
+      .select(window($"time", "9 seconds", "3 seconds", "2 second"), $"value")
+      .orderBy($"window.start".asc, $"value".desc).select("value")
+
+    Seq(df1, df2).foreach { df =>
+      val filter = df.queryExecution.optimizedPlan.find(_.isInstanceOf[Filter])
+      val exist = filter.get.constraints.iterator.toStream.filter(e =>
+        e.toString.contains(">=") || e.toString.contains("<"))
+      assert(exist.isEmpty, "No need to filter data between " +
+        "window.start and window.end when the sliding window length is not redundant")
+
+      checkAnswer(
+        df,
+        Seq(Row(4), Row(4), Row(4), Row(1), Row(1), Row(1), Row(2), Row(2), Row(2))
+      )
+    }
+
+    Seq(df3, df4).foreach { df =>
+      val filter = df.queryExecution.optimizedPlan.find(_.isInstanceOf[Filter])
+      val exist = filter.get.constraints.iterator.toStream.filter(e =>
+        e.toString.contains(">=") || e.toString.contains("<"))
+      assert(exist.isEmpty, "No need to filter data between " +
+        "window.start and window.end when the sliding window length is not redundant")
+
+      checkAnswer(
+        df,
+        Seq(Row(4), Row(4), Row(4), Row(1), Row(1), Row(1), Row(2), Row(2), Row(2))
+      )
+    }
+
+    // check produces right windows
```

Comment:
We don't need to validate the results here since other tests also do this. If you feel we don't have matching tests, please add separated general test.

Reply:
Most of the existing tests for sliding windows are under the condition of "`windowduration % slideduration ! = 0`". The only exception is the test case called "`millisecond precision sliding windows`", but it focuses on "`millisecond precision`". So I'm not sure if I need to add new case to verify that the change still produces right windows.

Reply:
Please add the case of `windowDuration % slideDuration == 0` in one of existing tests if you feel we missed the case. Let's focus to verify the actual change in new test.

Reply:
I think the test case called "millisecond precision sliding windows" has covered this situation.